functionary/builder/utils.py

```diff
@@ -33,34 +34,54 @@ def extract_package_definition(package_contents: bytes) -> dict:
 
     Returns:
         The package definition yaml loaded as a dict
+
+    Raises:
+        InvalidPackage: Raised whenever an error is encountered while extracting
+            package contents from the package.yaml in the tarball
     """
     package_contents_io = io.BytesIO(package_contents)
 
     try:
-        tarball = tarfile.open(fileobj=package_contents_io, mode="r")
-    except tarfile.ReadError:
+        tarfile = _get_tarfile(package_contents_io)
+        package_definition = _extract_package_definition(tarfile)
+        return package_definition
+    except ReadError:
         raise InvalidPackage(
             "Could not untar package file. Make sure it is a valid gzipped tarball."
         )
-
-    def close_files():
-        package_contents_io.close()
-        tarball.close()
-
-    try:
-        package_definition_io = tarball.extractfile("package.yaml")
     except KeyError:
-        close_files()
         raise InvalidPackage("package.yaml not found")
+    except (InvalidPackage, YAMLError):
+        raise InvalidPackage("package.yaml is invalid YAML")
+    finally:
+        tarfile.close()
```

Comment:
Something I only noticed when adding the view test is that this line will blow up if the `tarfile = _get_tarfile(package_contents_io)` is what throws the exception, since `tarfile` is not set in that case.

Reply:
Fixed